### ISSUE TITLE
redundant_pub_crate false positive

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust-clippy#8319]
### Summary

This code:

```rust
#![allow(dead_code, unused_variables)]
#![warn(clippy::nursery)]

mod foo {
    pub(crate) enum Bar<T, const LEN: usize> {
        Inline { len: u8, arr: [T; LEN] },
        Heap { v: Vec<T> }
    }

    impl <T: Copy + Default, const LEN: usize> Bar<T, LEN> {
        pub(crate) fn new() -> Self {
            Self::Inline {
                len: 0,
                arr: [T::default(); LEN],
            }
        }
    }
}

use foo::Bar;

fn main() {
    let sv = Bar::<u32, 10>::new();
}
```

Gives:

```rust
warning: pub(crate) enum inside private module
 --> src/main.rs:5:5
  |
5 |     pub(crate) enum Bar<T, const LEN: usize> {
  |     ----------^^^^^^^^^
  |     |
  |     help: consider using: `pub`
  |
note: the lint level is defined here
 --> src/main.rs:2:9
  |
2 | #![warn(clippy::nursery)]
  |         ^^^^^^^^^^^^^^^
  = note: `#[warn(clippy::redundant_pub_crate)]` implied by `#[warn(clippy::nursery)]`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_pub_crate
```

But if I replace "pub(crate) enum Bar" with "pub enum Bar" the code doesn't compile.

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:

**Comment by PatchMixolydic** [comment]
`pub enum Bar` seems to compile fine on 1.58 and nightly 1.60 ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=5e0f01dc2fa991504ecee933264dd965)).

**Comment by leonardo-m** [comment]
You're right, thank you. It's because I am usually compiling my code with a warning active (-W unreachable-pub):

```rust
warning: unreachable `pub` item
 --> ...\test.rs:5:5
  |
5 |     pub enum Bar<T, const LEN: usize> {
  |     ---^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |     |
  |     help: consider restricting its visibility: `pub(crate)`
  |
  = note: requested on the command line with `-W unreachable-pub`
  = help: or consider exporting it for use by other crates

warning: unreachable `pub` item
  --> ...\test.rs:11:9
   |
11 |         pub fn new() -> Self {
   |         ---^^^^^^^^^^^^^^^^^
   |         |
   |         help: consider restricting its visibility: `pub(crate)`
```
